pkg/ipcache/metadata.go

```diff
@@ -292,28 +325,85 @@ func RemoveLabelsFromIPs(
+		// toReplace stores identity pairs to replace in the ipcache.
+		toReplace = make(map[string]idPair)
 	)
+
+	IPIdentityCache.Lock()
+	defer IPIdentityCache.Unlock()
+
 	for prefix, lbls := range m {
-		id, exists := IPIdentityCache.LookupByIP(prefix)
+		id, exists := IPIdentityCache.LookupByIPRLocked(prefix)
 		if !exists {
 			continue
 		}
+
 		// Insert to propagate the updated set of labels after removal.
-		var la labels.LabelArray
-		if l := RemoveLabels(prefix, lbls, src); l != nil {
+		//
+		// There are two concurring cases here and they are annotated in the
+		// code below:
+		//   1) When labels are disassociated with an IP, the remaining set
+		//      either contains some labels or it's empty. In the former, the
+		//      identity is both marked for deletion and addition, because
+		//      there are labels still associated with the IP. In the latter,
+		//      the identity is marked for deletion only.
+		//   2) If removeLabels() indicates that a new identity needs to be
+		//      generated (newIdentity), a new identity is allocated (with the
+		//      appropriate labels removed), marked for addition, and lastly,
+		//      upserted into the ipcache (replacing the existing identity
+		//      entry). This can happen when an existing identity was *not*
+		//      released during removeLabels() due to the identity's refcount
+		//      being >1.
+		var (
+			newIdentity bool
+			la          labels.LabelArray
+		)
+		l, newIdentity := removeLabels(prefix, lbls, src)
+		if l != nil {
 			la = l.LabelArray()
 		}
 		idsToDelete[id.ID] = la
-		if len(lbls) > 0 {
+		if len(la) > 0 { // (1)
 			// If for example kube-apiserver label is removed from
-			// a remote-node, then RemoveLabels() will return a
+			// a remote-node, then removeLabels() will return a
 			// non-empty set representing the new full set of
 			// labels to associate with the node. In order to
 			// propagate the new identity, we must emit a delete
 			// event for the old identity and then an add event for
 			// the new identity.
 			idsToAdd[id.ID] = la
 		}
+		if newIdentity { // (2)
+			// If for example kube-apiserver label is removed from a CIDR
+			// identity (contains CIDR labels), then removeLabels() will return
+			// a non-empty set representing just the CIDR labels. We do the
+			// same as (1) from above, but additionally, we must also mark the
+			// new identity for addition and upsert the new identity into the
+			// IPCache.
+			newID, _, err := IdentityAllocator.AllocateIdentity(context.TODO(), l, false)
+			if err != nil {
+				log.WithError(err).WithFields(logrus.Fields{
+					logfields.IPAddr:         prefix,
+					logfields.Identity:       id,
+					logfields.IdentityLabels: l,    // new labels
+					logfields.Labels:         lbls, // removed labels
+				}).Error(
+					"Failed to allocate new identity after dissociating labels from existing identity. Traffic may be disrupted.",
+				)
+				continue
+			}
+			idsToAdd[newID.ID] = la
```

Comment:
I understand this, we will tell the policy selector layer below that `newID.ID` number points to the labels `la` and figure out which selectors select those labels, then cache the identity into there.

But then what confuses me is the `idsToAdd[id.ID] = la` case above. Could we end up associating the same set of labels to two different identities, and even change the meaning of an existing CIDR identity? This seems potentially a bit sketchy.

The only case I'm aware of where one identity should change its labels is the host identity case specific to this feature. In all other cases, Identities should be immutable.

Reply:
^^ The answer may be, the `idsToAdd[id.ID] = la` case above is not possible to trigger today, because there's only one source of identitymetadata labels. In that case, hmm maybe we can find some way to wrap this up nicely until we look at introducing further identitymetadata users, so that the next developer isn't surprised.

Reply:
> The only case I'm aware of where one identity should change its labels is the host identity case specific to this feature. In all other cases, Identities should be immutable.

Good point. The comments related to this are actually wrong as they mention `remote-node` instead of `local host`.

I think this is related to https://github.com/cilium/cilium/pull/18150#discussion_r771688052.

Reply:
https://github.com/cilium/cilium/pull/18150#discussion_r773565559